providers/implementations/ciphers/cipher_sm4_gcm.c

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* Dispatch functions for SM4 GCM mode */
+
+#include "cipher_sm4_gcm.h"
+#include "prov/implementations.h"
+#include "prov/providercommon.h"
+
+static void *sm4_gcm_newctx(void *provctx, size_t keybits)
+{
+    PROV_SM4_GCM_CTX *ctx;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    ctx = OPENSSL_zalloc(sizeof(*ctx));
+    if (ctx != NULL)
+        ossl_gcm_initctx(provctx, &ctx->base, keybits,
+                         ossl_prov_sm4_hw_gcm(keybits));
+    return ctx;
+}
+
+static OSSL_FUNC_cipher_freectx_fn sm4_gcm_freectx;
```

Comment:
This is being pre-declared yes `sm4_gcm_newctx` isn't.  That seems a little odd.
We generally make an attempt to include all of these pre-declarations in a block near the top of the file.  In this case, around line 15.

Keeping them in one place makes seeing what's implemented easier.  Having all of them gains type checking.

Reply:
It seems this kind of pre-declarations not that necessary. So I think either we have all functions pre-declared or we don't pre-declare any of them.

Reply:
This line is removed.

Reply:
I would much prefer all of these pre-declarations to be present.  If they aren't there, we will not detect problems with the function arguments/return value because of all the void casting used in the dispatch table.

This is a real problem.  I have already corrected multiple places where a function wasn't pre-declared and was incorrect.

Reply:
OK, then we may need to fix more like this.

Reply:
I've fixed all the ones I stumbled across doing other things but there are probably some left.
The worst were the ones where some of the functions were pre-declared but not all.